### ISSUE TITLE
Take pictures and upload to panoramax.mapcomplete.org

### ISSUE THREAD

**Issue by tordans** [bryceco/GoMap#822]
MapComplete always had a focus on collecting Pictures of OSM Objects. They used IMGUR until recently, when they switched to a special Panoramax Instance that is only used to handle pictures from the app.

Panoramax is an open source, federated Mapillary-like service.

More on the MapComplete news at https://en.osm.town/@MapComplete/113209991694821771

I think it would be a great addition to GoMap to allow storing geo referenced pictures.

- Use Case 1: Add pictures to OSM objects
   Users could take pictures, upload them to this instance and connect them in the data with a `panoramax=*` tag in one go.

- Use Case 2: OSM Notes
  A second use case would IMO be to allow picture uploads as part of the "add note" UI where I write a note like "check tags based on traffic sign (picture) and road markings (pictures)"

---

@pietervdvn would it be possible to allow GoMap (or StreetComplete) to upload pictures to your instance as well? What would the requirements be?

---

Ping https://github.com/pietervdvn/MapComplete/issues/1451

**Comment by pietervdvn** [comment]
It is @thibaultmol whom hosts the server and who thus has to decide this.

I don't see a reason why not, except the disk usage (but that shouldn't be a problem)

**Comment by thibaultmol** [comment]
I think it's probably best to keep separate instances for each app. To help with moderation and such. (it's just separate docker containers on the vm I have).
I don't mind looking into this yeah.

Did promise to put the main OSMBE instance up for the longest time, so probably going to prioritize that first unfortunately.

**Comment by mnalis** [comment]
- (ref: https://github.com/streetcomplete/StreetComplete/issues/5939)

**Comment by JLZIMMERMANN** [comment]
Such option would be really great.
I'm wishing a kind of photomapping object's that do not have any picture tag : wikimedia, picture, panoramax, mapillary ...
By this quest it would be a specific rendering for picture hunting for quality displaying.

Wikimedia did a specific website for this idea : [wikishootme ](https://wikishootme.toolforge.org/#lat=44.13648035933346&lng=4.809801578521729&zoom=17)
![image](https://github.com/user-attachments/assets/6603ea62-2b77-4bca-93a4-9ce0c06d48a6)

**Comment by bryceco** [comment]
A few questions:
- Why does MapComplete use its own instance of panoramax instead of using panoramax.openstreetmap.fr? 
- Does the user need to create an account on the panoramax instance before uploading? Is every photo associated with a user?
- Is there the expectation that users will be able to review and organize their photos inside the app?
- Who is responsible for content moderation?

**Comment by thibaultmol** [comment]
>     * Why does MapComplete use its own instance of panoramax instead of using panoramax.openstreetmap.fr?

Because the idea behind Panoramax is to spread out the servers so it's not centralized on one server.
 
>     * Does the user need to create an account on the panoramax instance before uploading? Is every photo associated with a user?

The way we did it with the Mapcomplete instance was the quick and dirty way of just having one user account and then using that login token for all uploads. In the actual image metadata it'll say Mapcomplete as it's author but besides that it will show the osm username. it's not ideal, but it was a quick way of implementing it. 
@Zverik has been working on a different solution so that the osm login you use for (in your case gomap) could be used to auth on the panoramax instance as well. for example.
 
>     * Is there the expectation that users will be able to review and organize their photos inside the app?

If you are referring to managing their photos on the panoramax instance? with our method: no. Because they're all on one user account. Again, not ideal.

With @zverik's approach they would be able to do that.

>     * Who is responsible for content moderation?

Me and pieter. Currently if an image gets reported, it automatically gets hidden. Until someone acts upon it. Currently moderation is kind of rough because there is no proper ui for it. But yeah

**Comment by bryceco** [comment]
> Is there the expectation that users will be able to review and organize their photos inside the app?

I meant inside of Go Map!! If a user takes a bunch of pictures and then wants to scroll through them, maybe delete some. Or maybe check if other people already uploaded a photo of the same object. This is probably something @tordans needs to answer.

**Comment by JLZIMMERMANN** [comment]
> * Is there the expectation that users will be able to review and organize their photos inside the app?

Panoramax is being improved with new goals : teams, organazations (such like Mapillary already doas), semantic added to pictures (IPTC + OSM tagging system) and you already may rename a sequence.

**Comment by bryceco** [comment]
I'm trying to figure out the scope of work. If we could just open a web page and have all the UI heavy lifting done by Panoramax then it makes things much easier and future proof.

**Comment by tordans** [comment]
Personally, those are the use cases I am hoping for:

### Taking a sequence of pictures

Something outside of GoMap. I ride my bike or go for a walk with the sole purpose of taking street level imagery. I use an app for that that is made for just this use case. Personally I review those before I upload to remove duplicates or weird/bad shots. (For long sequences there are multiple tools that help clean them up; I am [using this one](https://github.com/britiger/mapillary_gopro_tools).)
Those images belong in one of the main Panoramax instances, ideally of the given country.
I don't think GoMap or other apps should get involved with this.

### Taking pictures for notes

OSM Notes are great to discuss and shared things that are wrong or missing in OSM and pictures make them so much easier to work with. This is one of my GoMap Use Cases:
- Open an existing OSM note > adding 1-n pictures to it. The pic is uploaded to the special panoramax instance, because it is not a sequence and a different picture style; the app uploads and adds a comment with the URLs to the pictures. I consider the custom message "comment" optional in this flow.
- Create a new OSM note > same flow, but create the note and the comment should be required.

Complexity: Taking multiple pictures of one object / for one note will become complex fast. So maybe this should be scoped to just one pictures instead? One can always use this flow multiple times…

### Adding pictures to objects

Some objects in OSM really benefit from having a picture assigned to them. Some examples:
- Traffic signs
- Barriers on a cycle street, [like cycle barriers](https://wiki.openstreetmap.org/wiki/Tag:barrier%3Dcycle_barrier)
- Art/Statues/…
- Maybe `hazard=*` things

(The [related tags list](https://taginfo.openstreetmap.org/keys/mapillary#combinations), [2](https://taginfo.openstreetmap.org/keys/panoramax#combinations) is not helpful to find other examples, unfortunatelly)

This is the second UseCase I would love to see in GoMap:
- On a given OSM object > At the end of the preset/tag list, use a picture button > Take a picture, upload it, add the tag to GoMap

Note: The tagging that is used for this by MapComplete is `panoaramx=ID,` `panoramax:1=ID` to work around value field size limitations. More at https://github.com/ideditor/schema-builder/issues/180

### Looking at pictures on elements

With both cases – notes and OSM objects – we now need a way to look at the picture.
Personally I don't see myself using GoMap to look at many pictures, that is for other workflows that likely start at home in iD or some QA tool.

But some basic way to 
- see that there is a picture added
- have a way to look at it full screen
- have a way to open the source URL at panoramax (this requires something that resolves the right instance or takes the meta instance every time?)
- have a way to download the image

UI:
- For notes, that could be a preview of the image in the UI with tab to see details.
- For OSM objects, that could be a tiny preview next to the input field which can be tapped.

BUT, all this can become a lot of UI and the minimal thing would be to just open a web browser in the panoramax instace.
- Notes: There is a link in the note already: That could oben an embedded browser and be done with it
- OSM Objects: That would need a custom link icon next to the input like we have in other cases.

### Looking at pictures on the map

Personally, I don't thing this an important use case.
Again, I don't see myself working with the pictures in GoMap but in other tools.

The UseCase I see here would be:
- Use the new "task" layer to hint to me where pictures are missing for use cases where I want to collect them.
- Maybe there is a task type "picture" that will take me to the camera UI right from the task-map-pin, so I save a few clicks.

I could see myself adding tasks for all of the objects I listed above to trigger the "completeionist gene" and add pictures to all of them in my hood.


### UseCases I don't see
- Deleted pictures that I uploaded I already uploaded – I will have to log into the panoramax instance to do that or use the "delete that" links that the instances provides
- Working with the pictures in GoMap – see above


---

> If we could just open a web page and have all the UI heavy lifting done by Panoramax then it makes things much easier and future proof.

I am not sure if this will work nicely. The UI flow will require a back and forth which I image will be hard to implement with a web based process? How would the app get the panoramax-ID after the upload was done?

**Comment by bryceco** [comment]
> How would the app get the panoramax-ID after the upload was done?

You're correct. I guess the photo-taking and upload would need to be handled app-side, but other things (viewing photos, deleting photos, etc) would be done via web page.

**Comment by JLZIMMERMANN** [comment]
In my dream usecase : a photo mapping hunting option (like treasure game). A specific layer to go anywere and show main objects that haven't yet any picture.

The first thet would be shown would be tourism=* and important amenity=*...

**Comment by bryceco** [comment]
First cut of uploading and viewing panoramax photos is in the most recent beta release.

**Comment by tordans** [comment]
@bryceco thanks a lot for adding this feature!

I tested it a bit – here is some unordered feedback.

---

## Test 1: Mapcomplete => GoMap

Uploaded images with MapComplete in https://www.openstreetmap.org/changeset/164132059
- Element https://www.openstreetmap.org/way/297142651 and 
- Element https://www.openstreetmap.org/node/43662888

One of the images:
- Viewer https://api.panoramax.xyz/#focus=meta&map=17/51.957068/7.617782&pic=9435e8df-a13f-4276-b320-880a8e1b7585&speed=250&xyz=255.00/0.00/30
- API https://api.panoramax.xyz/api/collections/6e702976-580b-419c-8fb3-cf7bd364e6f8/items/9435e8df-a13f-4276-b320-880a8e1b7585
        `"geovisio:image": "https://panoramax.mapcomplete.org/permanent/94/35/e8/df/a13f-4276-b320-880a8e1b7585.jpg",`

## Test 2: Upload via GoMap

- Element https://www.openstreetmap.org/node/389872045
- Viewer https://api.panoramax.xyz/#focus=meta&map=17/51.965351/7.615701&pic=f57f084b-82ae-4472-b013-d2ec4f8e2a4f&speed=250&xyz=0.00/0.00/30
- API https://api.panoramax.xyz/api/collections/cd5507e1-d34f-47c7-9ff5-500347b76517/items/f57f084b-82ae-4472-b013-d2ec4f8e2a4f
         ` "geovisio:image": "https://panoramax.openstreetmap.fr/images/f5/7f/08/4b/82ae-4472-b013-d2ec4f8e2a4f.jpg",`

----

- When I took the image with GoMap, I had to OAuth somewhere. I did not look properly where but the flow did not explain what was happening and why. It all worked but left me wondering what I just did. It's all very small and when on the go it's hard to understand what is happening. I am assuming I did an OAuth with OSM against Panoramax FR? Is that stored in GoMap or will I need to do this again next time? — Maybe there is a way to give some more Kontext before this happens the first time?

- <img width="361" alt="Image" src="https://github.com/user-attachments/assets/7a18a17e-2584-4214-b659-658b3e42b1f7" />
  - That icon is a bug, right? Is it supposed to be two icons or one?
  - I think the "no image yet" case should start the "new image" flow right away and not open the panel first

- ![Image](https://github.com/user-attachments/assets/935ca2d3-7172-45c9-ad14-7deee3aa6bcb)
     <img width="676" alt="Image" src="https://github.com/user-attachments/assets/f504e1bf-34ce-4fdd-af2b-1bf7f12bbfbc" />
   - What does the blue logo at the top do? I see a Panoramax Map of europe but don't know what to do with that page. Can I choose an image from there to add an existing image?
   - I think in the "no image there yet" case, there should not be a preview placeholder image but a big button to start the new image flow


- ![Image](https://github.com/user-attachments/assets/e99f926e-628d-412e-9687-8b6e11c6c72a) 
     ![Image](https://github.com/user-attachments/assets/cad17372-d459-4043-b996-2c4c2ff89480)
    - The uploads are done to osm-fr right? I think it would be great if the apps came together and used the same instance for those non-sequence shots. The last I read is that MapComplete does not want to share it's instance. But maybe this can be talked about again…?
   - Looking at images that where uploaded via MapComplete fails AFAIK and I think it happens because you use the `fr` and not the `xyz` domain. My understanding is that the `xyz` domain is the (only) meta-server.

- ![Image](https://github.com/user-attachments/assets/34662111-44e9-4798-9bf4-d104b71f6ef6)
   - For MapComplete, it would be ideal to optionally show the `Exif.Image.Artist` as well. That would be what is visible on the website from MapComplete and also the Viewer
      - [Viewer](https://api.panoramax.xyz/#focus=meta&map=17/51.957068/7.617782&pic=9435e8df-a13f-4276-b320-880a8e1b7585&speed=250&xyz=255.00/0.00/30) <img width="348" alt="Image" src="https://github.com/user-attachments/assets/220e05c9-c7a1-4fba-8ab9-2b94ac25ad83" />

- Does the preview load the big image or the preview. The API calls above mention a smaller version. And the load time is rather bad for me, so maybe there is a smaller preview?

- Where does GoMap explain to users what is happening with this feature? For example, that images are public (get anonymized) and shared under a specific license etc.

**Comment by bryceco** [comment]
@tordans 

> When I took the image with GoMap, I had to OAuth somewhere.

Yes, I'm uploading images to panoramax.openstreetmap.fr as you guessed, which authenticates against the OSM server. MapComplete uses a private instance with a single user account that shares all uploads, so authentication isn't required. The advantage of having photos tied to an OSM account is:
* Users can manage their own photos on the panoramax website. They can delete photos, etc.
* If a user is uploading inappropriate material its easy to identify and ban them.

The credentials are stored in cookies as part of your regular Safari profile, so if you visit panoramax.openstreetmap.fr in Safari before or after uploading then you're already authenticated, and the credentials will persist for several weeks or until you clear cookies.

> That icon is a bug, right? Is it supposed to be two icons or one?

That is Apple's camera icon.

> What does the blue logo at the top do? I see a Panoramax Map of europe but don't know what to do with that page. Can I choose an image from there to add an existing image?

You can't do anything interesting on that page, but you can use it to navigate to your personal profile where your image uploads are stored.  We should probably change the URL to "https://panoramax.openstreetmap.fr/your-pictures"

> Looking at images that where uploaded via MapComplete cannot be viewed AFAIK and I think it happens because you use the fr and not the xyz domain. 

When trying to download images I try both the .xyz and the .fr domains, because uploaded images appear first in .fr and later in .xyz. There are various delays so if you upload from MapComplete it may not show up in xyz for a while (up to an hour in my limited experience).

> For MapComplete, it would be ideal to optionally show the Exif.Image.Artist as well. 

You mean when using MapComplete as a viewer? Since it's already tied to an OSM ID MapComplete should be able to extract that. 

> Does the preview load the big image or the preview. 

I'm downloading the "sd.jpg" version rather than the "hd.jpg" version, so it should be smaller (but it's still large).

> Where does GoMap explain to users what is happening with this feature? For example, that images are public (get anonymized) and shared under a specific license etc.

It doesn't! 😅 I guess we could add some text to the initial window for that.

> Can you share a quick note on how you implemented this from an Auth point of view 

I open a Safari window at https://panoramax.openstreetmap.fr/api/auth/login, and provide client_id and scope info similar to doing an OAuth2 handshake. The panoramax server contacts the OSM server and pops up its own OAuth2 page that the user agrees to, and then finally it sends us a cookie containing an authentication token. That cookie/token is automatically provided to the server when we do additional calls.

> It would be great to have the same feature for OSM Notes. To be able to add an image URL to the note right from the new note or new note comment UI

Agreed, but that is more confusing from a user perspective compared to clicking a field named "panoramax"
